src/sidecar/websocketManager.ts

```diff
@@ -0,0 +1,335 @@
+        logger.debug(
+          `Recieved ${headers.message_type} websocket message from originator ${headers.originator}: ${JSON.stringify(message, null, 2)}`,
+        );
+
+        // Defer to the NodeJS EventEmitter to deliver the message to the registered by-message-type async handler(s).
+        this.deliverToCallbacks(message).catch((e) => {
+          logger.error(`Error delivering message ${JSON.stringify(message, null, 2)}: ${e}`);
+        });
+      });
+
+      this.disposables.push({
+        dispose: () => {
+          // Close the websocket when we're disposed of.
+          if (this.websocket && this.websocket.readyState !== WebSocket.CLOSED) {
+            this.websocket.close();
+            this.websocket = null;
+          }
+        },
+      });
+    });
+  }
+
+  /**
+   * Send a message to / through sidecar over the websocket.
+   * The websocket send is ultimately async underneath the hood.
+   * @throws {WebsocketClosedError} if the websocket is not connected.
+   */
+  public send<T extends MessageType>(
+    message: Message<T>,
+    websocket: WebSocket | undefined = undefined,
+  ): void {
+    if (!websocket) {
+      if (!this.websocket) {
+        logger.error("Websocket not assigned or provided, cannot send message");
+        throw new WebsocketClosedError();
+      }
+      websocket = this.websocket;
+    }
+
+    if (websocket && websocket.readyState === WebSocket.OPEN) {
+      const payload = JSON.stringify(message);
+      if (payload.length > 64 * 1024) {
+        const errorMessage = `Cannot send websocket message, too large: ${payload.length} bytes`;
+        logger.error(errorMessage);
+        window.showErrorMessage(errorMessage);
+        throw new Error(errorMessage);
+      }
+      logger.debug(`Sending ${payload.length} byte ${message.headers.message_type} message`);
+      websocket.send(payload);
+    } else {
+      logger.error("Websocket not assigned + open, cannot send message");
+      throw new WebsocketClosedError();
+    }
+  }
+
+  /** How many peer workspaces are connected to sidecar, exclusive of ourselves? */
+  public getPeerWorkspaceCount(): number {
+    return this.peerWorkspaceCount;
+  }
+
+  // Subsciption and delivery of messages to the appropriate async handlers
+
+  /**
+   * Register an async callback for messages of the given type. The callback
+   * will be called by `deliverToCallbacks()` and awaited every time a message of the given type is delivered.
+   **/
+  public subscribe<T extends MessageType>(messageType: T, callback: MessageCallback<T>): void {
+    this.messageRouter.on(messageType, callback);
+  }
+
+  /**
+   * Register an async callback for messages of the given type, to be called only once by
+   * `deliverToCallbacks` upon  delivery of the next message of that type.
+   **/
+  public once<T extends MessageType>(messageType: T, callback: MessageCallback<T>): void {
+    this.messageRouter.once(messageType, callback);
+  }
```

Comment:
@jlrobins are we still planning to get rid of these, give that Node's EE has generic type?

Reply:
Not planning to get rid of per se, in that I don't want to publicly expose the implementation of message delivery, so will still have to have some methods for registration.